xds/src/main/java/io/grpc/xds/WeightedTargetLoadBalancer.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.XdsSubchannelPickers.BUFFER_PICKER;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.ConnectivityState;
+import io.grpc.InternalLogId;
+import io.grpc.LoadBalancer;
+import io.grpc.Status;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.WeightedRandomPicker.WeightedChildPicker;
+import io.grpc.xds.WeightedRandomPicker.WeightedPickerFactory;
+import io.grpc.xds.WeightedTargetLoadBalancerProvider.WeightedPolicySelection;
+import io.grpc.xds.WeightedTargetLoadBalancerProvider.WeightedTargetConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/** Load balancer for weighted_target policy. */
+final class WeightedTargetLoadBalancer extends LoadBalancer {
+
+  private final XdsLogger logger;
+  private final Map<String, GracefulSwitchLoadBalancer> childBalancers = new HashMap<>();
+  private final Map<String, ChildHelper> childHelpers = new HashMap<>();
+  private final Helper helper;
+  private final WeightedPickerFactory weightedPickerFactory;
+
+  private Map<String, WeightedPolicySelection> targets = ImmutableMap.of();
+
+  WeightedTargetLoadBalancer(Helper helper) {
+    this(
+        checkNotNull(helper, "helper"),
+        WeightedRandomPicker.RANDOM_PICKER_FACTORY);
```

Comment:
This looks bad... You should rather call constructor to create a factory or call the factory class to get the instance. At least, not something that is not the factory itself.